File: leetcode/0143-reorder-list.cc
// https://leetcode.com/problems/reorder-list/
// Medium
// Mock

#include<vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head) return;
        vector<ListNode*> v;
        ListNode* tempHead = head;
        int count = 0;
        while (tempHead != nullptr) {
            v.push_back(tempHead);
            tempHead = tempHead->next;
            ++count;
        }
        ListNode* next = head->next;
        ListNode* temp = nullptr;
        for (int i = 0; i < count/2; ++i) {
            temp = v.back();
            v.pop_back();
            head->next = temp;
            if (temp == next) {
                temp->next = nullptr;
                return;
            }
            temp->next = next;
            head = next;
            next = head->next;
        }
        head->next = nullptr;
        return;
    }
};

File: leetcode/python/0143-reorder-list.py
# https://leetcode.com/problems/reorder-list/
# Medium

# This is the O(1) memory solution
# 1. Use fast and slow pointer to find where the middle is
# 2. Reverse the second half of the linked list
# 3. Merge the two lists that you have
#    - Merge as 'while prev and prev.next'
#      because we do not want loops
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        curr, prev = slow, None
        while curr:
            curr.next, curr, prev = prev, curr.next, curr
        
        curr = head
        while prev and prev.next:
            curr.next, curr = prev, curr.next
            prev.next, prev = curr, prev.next
            
        return head

# Use a list to keep track of all the nodes
# Keep a 'curr' node and then pop from the list, and 
#   set curr.next to the pop as well as viceversa
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        curr = head
        l = []
        while curr:
            l.append(curr)
            curr = curr.next

        curr = head
        right = l.pop()
        while curr != right:
            temp = curr.next
            curr.next = right
            curr = temp
            if curr == right:
                break
            right.next = curr
            right = l.pop()
        curr.next = None
        return head